### ISSUE TITLE
Mipha-3B shows unexpectedly low accuracy on ChartQA — is my evaluation code incorrect?

### ISSUE THREAD

**Issue by shenlayu** [zhuyiche/llava-phi#26]
Hi team, thanks for the great work on Llava-phi and Mipha! I’m currently evaluating Mipha-3B on the ChartQA benchmark (https://github.com/vis-nlp/ChartQA), but I’m observing very poor performance — significantly lower than expected (<20% EM accuracy). I’m wondering if there’s something wrong with how I’m using the model or constructing prompts. We would typically expect 60–80% EM for chart reasoning tasks on fine-tuned VLMs, so I’m wondering if this is a prompt mismatch or decoding issue.

📉 Observation

The model seems to respond with verbose, sometimes hallucinated answers instead of concise numerical/text responses that match the expected ChartQA style. For instance, instead of returning just 17, it returns a sentence like:

Question: What is the median value? Answer with only a number, a name from the chart, or yes/no. Do not explain.
Answer: No
GT: 17

Question: What was the number of patients waiting on trolleys in August 2019? Answer with only a number, a name from the chart, or yes/no. Do not explain.
Answer: No patients waiting on trolleys
GT: 9562

By the way, more than 50% of questions are answered with yes/no although their true answers are a number or noun.

📦 Inference script used

Below is the full inference script I used for ChartQA evaluation:

```python
import argparse
import torch
import os
import json
from tqdm import tqdm

from mipha.constants import IMAGE_TOKEN_INDEX, DEFAULT_IMAGE_TOKEN, DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN
from mipha.conversation import conv_templates, SeparatorStyle
from mipha.model.builder import load_pretrained_model
from mipha.utils import disable_torch_init
from mipha.mm_utils import tokenizer_image_token, get_model_name_from_path, KeywordsStoppingCriteria, process_images

from torch.utils.data import Dataset, DataLoader
from PIL import Image
import math


def split_list(lst, n):
    return [lst[i:i + math.ceil(len(lst) / n)] for i in range(0, len(lst), math.ceil(len(lst) / n))]


def get_chunk(lst, n, k):
    return split_list(lst, n)[k]


class ChartQADataset(Dataset):
    def __init__(self, items, image_folder, tokenizer, image_processor, model_config, conv_mode):
        self.items = items
        self.image_folder = image_folder
        self.tokenizer = tokenizer
        self.image_processor = image_processor
        self.model_config = model_config
        self.conv_mode = conv_mode

    def __len__(self):
        return len(self.items)

    def __getitem__(self, idx):
        item = self.items[idx]
        qid = item["id"]
        image_path = os.path.join(self.image_folder, item["image"])
        raw_q = item["conversations"][0]["value"].replace("<image>", "").strip()
        suffix = " Answer with only a number, a name from the chart, or yes/no. Do not explain."
        question = raw_q + suffix

        if self.model_config.mm_use_im_start_end:
            question_with_image = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + '\n' + question
        else:
            question_with_image = DEFAULT_IMAGE_TOKEN + '\n' + question

        conv = conv_templates[self.conv_mode].copy()
        conv.append_message(conv.roles[0], question_with_image)
        conv.append_message(conv.roles[1], None)
        prompt = conv.get_prompt()

        input_ids = tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt')
        input_ids = input_ids.unsqueeze(0)

        image = Image.open(image_path).convert('RGB')
        image_tensor = self.image_processor.preprocess(image, return_tensors='pt')['pixel_values'][0].unsqueeze(0)

        gt_answer = item["conversations"][1]["value"].strip()

        return {
            "qid": qid,
            "question": question_with_image,
            "input_ids": input_ids,
            "image_tensor": image_tensor,
            "gt_answer": gt_answer
        }


def eval_chartqa(args):
    disable_torch_init()

    # Load model
    model_path = os.path.expanduser(args.model_path)
    model_name = get_model_name_from_path(model_path)
    tokenizer, model, image_processor, context_len = load_pretrained_model(
        model_path, args.model_base, model_name
    )

    # Load data
    with open(args.question_file, "r") as f:
        raw_data = json.load(f)

    chunked_data = get_chunk(raw_data, args.num_chunks, args.chunk_idx)

    dataset = ChartQADataset(
        chunked_data, args.image_folder, tokenizer, image_processor, model.config, args.conv_mode
    )
    dataloader = DataLoader(dataset, batch_size=1, shuffle=False)

    model.eval()
    model.cuda()

    stop_str = conv_templates[args.conv_mode].sep2
    stop_criteria = lambda ids: [KeywordsStoppingCriteria([stop_str], tokenizer, ids)]

    results = []

    for batch in tqdm(dataloader):
        input_ids = batch["input_ids"].squeeze(0).cuda()
        image_tensor = batch["image_tensor"].squeeze(0).half().cuda()
        qid = batch["qid"][0]
        question = batch["question"][0]
        gt_answer = batch["gt_answer"]

        with torch.inference_mode():
            output_ids = model.generate(
                input_ids=input_ids,
                images=image_tensor.unsqueeze(0),
                do_sample=True if args.temperature > 0 else False,
                temperature=args.temperature,
                top_p=args.top_p,
                num_beams=args.num_beams,
                max_new_tokens=128,
                use_cache=True,
                pad_token_id=tokenizer.eos_token_id,
                eos_token_id=tokenizer.eos_token_id,
                stopping_criteria=stop_criteria(input_ids),
            )

        input_len = input_ids.shape[1]
        output_text = tokenizer.batch_decode(output_ids[:, input_len:], skip_special_tokens=True)[0].strip()
        if output_text.endswith(stop_str):
            output_text = output_text[:-len(stop_str)].strip()

        results.append({
            "id": qid,
            "question": question,
            "pred": output_text,
            "gt": gt_answer
        })

    os.makedirs(os.path.dirname(args.answers_file), exist_ok=True)
    with open(args.answers_file, "w") as fout:
        json.dump(results, fout, indent=2)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model-path", type=str, required=True)
    parser.add_argument("--model-base", type=str, default=None)
    parser.add_argument("--image-folder", type=str, required=True)
    parser.add_argument("--question-file", type=str, required=True)
    parser.add_argument("--answers-file", type=str, required=True)
    parser.add_argument("--conv-mode", type=str, default="v0")
    parser.add_argument("--num-chunks", type=int, default=1)
    parser.add_argument("--chunk-idx", type=int, default=0)
    parser.add_argument("--temperature", type=float, default=0.2)
    parser.add_argument("--top_p", type=float, default=None)
    parser.add_argument("--num_beams", type=int, default=1)
    args = parser.parse_args()

    eval_chartqa(args)
```

⸻

🧩 Questions
	1.	Is the above inference code compatible with how Mipha-3B expects prompts for ChartQA-like tasks?
	2.	Am I missing any postprocessing step (e.g., answer extraction, prompt format, conversation template)?
	3.	Has Mipha-3B been evaluated on ChartQA officially? If not, are there recommended prompt modes or datasets for quantitative evaluation?

⸻

Any clarification or suggestions would be hugely appreciated!

Thanks 🙏

**Comment by shenlayu** [comment]
Btw, this code snippet at `mipha/model/builder.py:74` is really confusing to me
```python
if 'phi' or "gemma" in model_name.lower():
```
